soc/arm/atmel_sam0/common/bossa.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2020 Google LLC.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <drivers/usb/usb_dc.h>
+#include <usb/class/usb_cdc.h>
+
+/*
+ * Magic value that causes the bootloader to stay in bootloader mode instead of
+ * starting the application.
+ */
+#if CONFIG_BOOTLOADER_BOSSA_ADAFRUIT_UF2
+#define DOUBLE_TAP_MAGIC 0xf01669ef
+#elif CONFIG_BOOTLOADER_BOSSA_ARDUINO
+#define DOUBLE_TAP_MAGIC 0x07738135
+#else
+#error Unsupported BOSSA bootloader variant
+#endif
+
+#if defined(CONFIG_CDC_ACM_ARDUINO_RESET_SUPPORT)
+
+static void bossa_reset(struct device *dev)
+{
+	uint32_t *top;
+
+	/* The programmer set the baud rate to 1200 baud.  Reset into the
+	 * bootloader.
+	 */
+	usb_dc_detach();
+
+	top = (uint32_t *)(DT_REG_ADDR(DT_NODELABEL(sram0)) +
+			   DT_REG_SIZE(DT_NODELABEL(sram0)));
+	top[-1] = DOUBLE_TAP_MAGIC;
+
+	NVIC_SystemReset();
+}
+
+static int bossa_init(struct device *unused)
+{
+	struct device *dev =
+		device_get_binding(CONFIG_USB_CDC_ACM_DEVICE_NAME "_0");
```

Comment:
How do you know this is the right device to use?  Possibly replace by `CONFIG_UART_CONSOLE_ON_DEV_NAME` as there doesn't appear to be any coupling between devicetree identifiers and USB devices.

Then check whether the reset callback succeeds.

Reply:
The console alias is no good, as it's common to have the console on a physical UART.

Leaving open while I think about it.

Reply:
How about a Kconfig `BOSSA_UART_LABEL`, defaulted to `"CDC_ACM_0"`?   (That the label is always for a CDC-ACM UART doesn't change the fact that from the perspective of the protocol it's a UART, though the limitation could be noted in the help section of the option.)

Reply:
I went for `_DEVICE_NAME` to match the style of `CONFIG_USB_CDC_DEVICE_NAME`